### ISSUE TITLE
Integration with deep learning frameworks

### ISSUE THREAD

**Issue by lesteve** [dask/dask-ml#268]
Collecting related thoughts from https://github.com/dask/dask-ml/issues/210.

@stsievert in https://github.com/dask/dask-ml/issues/210#issuecomment-401836989

> Especially with PyTorch. It certainly feels like there should be an integration with PyTorch because it has [torch.distributed](https://pytorch.org/docs/stable/distributed.html) and [torch.multiprocessing](https://pytorch.org/docs/stable/multiprocessing.html)

@stsievert in https://github.com/dask/dask-ml/issues/210#issuecomment-397023505
> Optimization:
> * I see these algs (e.g., Hogwild) as exploiting dask's distributed architecture
> * These will require a parameter server. Can we make this general and integrate with (for example) CuPy/Chainer and PyTorch?

@TomAugspurger in https://github.com/dask/dask-ml/issues/210#issuecomment-397005038

> Dask-Tensorflow
> - Review new datasets API, anything we should do there?

**Comment by TomAugspurger** [comment]
Thanks @lesteve!

I'm relatively unfamiliar with these libraries, especially with their distributed runtimes, where Dask may be most useful.

I'd be curious to hear from people who have experience here.

**Comment by mrocklin** [comment]
I'm also unfamiliar with using any modern deep learning library.  I'd love to hear from people who do use them, want to use them in a distributed way, and have experienced some pain while trying to do so.  (also cc'ing @bnaul @seibert)

My current understanding is that people do the following:

1.  Build up a model.  For this they use Tensorflow, Keras, PyTorch, etc..  I don't think Dask has any role to play here
2.  Choose an optimizer, like AdaGrad.  Dask might play a role here, or it might modify this optimizer, similar to the approach taken in [Hovorod](https://github.com/uber/horovod)

    ```python
    # Build model...
    loss = ...
    opt = tf.train.AdagradOptimizer(0.01 * hvd.size())

    # Add Horovod Distributed Optimizer
    opt = hvd.DistributedOptimizer(opt)
    ```
3.  Load and preprocess data.  Numpy/Pandas is probably active here, maybe Dask as well
4.  Hand that data to the model/optimizer for training

**Comment by mrocklin** [comment]
Personally, I'm curious to see what workflows that might potentially use Dask and a deep learning framework would look like.  That is something productive that people can do now that might help to focus the discussion.

**Comment by lesteve** [comment]
I guess this is not really integration per se so I opened a different issue #281 about the use case that some people are trying to tackle by combining dask and deep learning frameworks.

**Comment by mrocklin** [comment]
Questions like that are welcome.  It's nice to identify issues that come up
in practice, even if they are less research-y.

**Comment by arthurmensch** [comment]
Currently `pytorch` only documented way of doing Hogwild training is through using `Process()` commands and inheritance of models. See for instance https://github.com/pytorch/examples/blob/master/mnist_hogwild/main.py.

 This is not compatible with the modern approach of using dask pools, or even multiprocess pools. I don't really see a simple way around that.

**Comment by corrado9999** [comment]
My professional experience in which I believe daks-ml would be great.

I work with remote sensing data. If you are thinking of Google Maps that is only part of the story. Going to the point, we have images of several gigabytes (tens of thousands of pixel, tens of channels) acquired tens times each year. Thanks to NASA and ESA (European Space Agency), we have 4 great pools of images available for free: MODIS, LANDSAT-8, Sentinel-1 and Sentinel-2.

Now, I am facing the issue to train some deep network on these data. They are all available on AWS (https://registry.opendata.aws/?search=satellite%20imagery), so it makes perfect sense to avoid the downloading and train the model directly on the cloud. As features we use bands from one (or more!) datasource, but also time must be taken into account in some way (this not something dask can help with, but it could somehow affect the design).

The training is usually supervised, the target being classification (water, urban, grassland, cropland, ...). We are also interested in 3 main kind of classification:
1. Pure pixel based: each pixel is assigned to one class, without knowing where it is located
1. Pixel based with neighborhood: this is where CNN come to play, because we want to exploit information coming from near pixel in order to estimate some spatial features
1. Object based: we aggregate pixels based on some vector data (known _a priori_)

Please, take into account that we have a lot of data, and that data storage usually costs more than processing. So, in some cases it may be preferable to perform data preprocessing/normalization at each batch, in some other cases we may prefer to cache them (instance based). For the same reason, along and across epochs we would like to minimize data movement, but yet mix batches coming from different images in most possible ways.

Sorry for the long post, I hope to have been clear enough about my use case.

**Comment by TomAugspurger** [comment]
Thanks Joseph. Based on your description,
https://github.com/dask/dask-examples/issues/35 sounds quite similar to
your workflow. It'd be nice if we could develop that into a fully-formed
example.

**Comment by stsievert** [comment]
https://github.com/dask/distributed/issues/2581 is on training a PyTorch model with Dask.

**Comment by AlbertDeFusco** [comment]
I've made progress using a Dask DataFrame with the Keras `.fit_generator()` method to mimic `Incremental` for Scikit-Learn. 

https://anaconda.org/defusco/keras-dask/notebook

I'm close to getting `dask_ml.wrappers.ParallelPostfit`, and `keras.wrappers.scikit_learn.KerasClassifier()` working with sklearn piplines.

**Comment by TomAugspurger** [comment]
Thanks Albert, looks interesting. I think providing a DaskGenerator, or at
least documentation on how to write one, would be very useful.

A couple questions

1. With your DaskGenerator, if I do `gen = DaskGenerator(X, y); gen[0];
gen[1]`, then I *think* that you'll end up redoing a bunch of computation.
Does that sound correct? In this example I think you would end up refitting
the entire StandardScaler, splitting, etc. The solution is to persist the
transformed data, either in memory on the cluster, perhaps as part of
DaskGenerator.__init__, or on disk. And then the data passed to
DaskGenerator would be loaded from disk. (LMK if I'm not making sense)
2. Are you using distributed at all? It doesn't look like it. I ask because
I've never gotten Keras / Tensorflow to work properly in multiple processes.

**Comment by AlbertDeFusco** [comment]
@TomAugspurger ,

1. Yes, it would be good to persist (if possible) any data given to `DaskGenerator`.

2. No, I have not attempted this with Distributed.

**Comment by CMCDragonkai** [comment]
I was thinking of using Dask-ML to replace the multi-gpu model that exists Keras. So than we have a dask worker per-GPU instead of per-core as it is usually.

**Comment by lesteve** [comment]
Not sure what you are trying to do exactly and I am not an expert on Keras so I just want to comment on this part:

> So than we have a dask worker per-GPU instead of per-core as it is usually.

There is a way to have a dask-worker per-GPU setting CUDA_VISIBLE_DEVICES environment variable in the worker processes and people have been doing it for some time (for example here is a Scipy 2016 [video](https://youtu.be/jLQ2wAwapRg?t=914) about combining dask and numba). A more recent approach is https://github.com/rapidsai/dask-cuda.

**Comment by CMCDragonkai** [comment]
Hmm, it's a bit more complicated than that. We would also need a parameter server in the CPU land so that it accumulates weight updates from all the models. But that would make `dask-ml` a full ML framework akin to Keras, and instead of going dask -> keras -> tensorflow, it would be dask -> tensorflow.

**Comment by skeller88** [comment]
> My professional experience in which I believe daks-ml would be great.
> 
> I work with remote sensing data. If you are thinking of Google Maps that is only part of the story. Going to the point, we have images of several gigabytes (tens of thousands of pixel, tens of channels) acquired tens times each year. Thanks to NASA and ESA (European Space Agency), we have 4 great pools of images available for free: MODIS, LANDSAT-8, Sentinel-1 and Sentinel-2.
> 
> Now, I am facing the issue to train some deep network on these data. They are all available on AWS (https://registry.opendata.aws/?search=satellite%20imagery), so it makes perfect sense to avoid the downloading and train the model directly on the cloud. As features we use bands from one (or more!) datasource, but also time must be taken into account in some way (this not something dask can help with, but it could somehow affect the design).
> 
> The training is usually supervised, the target being classification (water, urban, grassland, cropland, ...). We are also interested in 3 main kind of classification:
> 
> 1. Pure pixel based: each pixel is assigned to one class, without knowing where it is located
> 2. Pixel based with neighborhood: this is where CNN come to play, because we want to exploit information coming from near pixel in order to estimate some spatial features
> 3. Object based: we aggregate pixels based on some vector data (known _a priori_)
> 
> Please, take into account that we have a lot of data, and that data storage usually costs more than processing. So, in some cases it may be preferable to perform data preprocessing/normalization at each batch, in some other cases we may prefer to cache them (instance based). For the same reason, along and across epochs we would like to minimize data movement, but yet mix batches coming from different images in most possible ways.
> 
> Sorry for the long post, I hope to have been clear enough about my use case.

@corrado9999 , have you developed a good ml data preparation workflow since your question? I have a similar project I'm working on. 

The problem I see is that batches are generated randomly, and so that requires random selections of the dataset. As long as the dataset is not in memory, selections will be slow.  But I think https://www.tensorflow.org/guide/data_performance provides ways around that with prefetching.

**Comment by corrado9999** [comment]
@skeller88 , no, unfortunately I have not worked on this (yet).

> The problem I see is that batches are generated randomly, and so that requires random selections of the dataset. As long as the dataset is not in memory, selections will be slow. 

Sure we need a pipeline, in order to perform such slow actions while the model is training. The TF API you pointed out is actually very interesting, not sure how it could be integrated in this framework though.

**Comment by skeller88** [comment]
Gotcha. I ended up using the tensorflow datasets API directly. It has batch prefetching, which makes performance a bit better. See my code [here](https://github.com/skeller88/deep_learning_project/blob/master/data_science/keras/dataset.py).

**Comment by stsievert** [comment]
One straightforward integration would be hooking in Tensorboards HParams dashboard to visualize how well different parameters performed. Here’s a description of basic usage: https://www.tensorflow.org/tensorboard/hyperparameter_tuning_with_hparams

I think this would amount to writing logs in a specific format. Here’s how to get some example logs, and run tensorboard on those logs (code pulled from the post above):


``` bash
$ wget -q 'https://storage.googleapis.com/download.tensorflow.org/tensorboard/hparams_demo_logs.zip'
$ unzip -q hparams_demo_logs.zip -d logs/hparam_demo
$ tensorboard --logdir logs/hparam_demo
```

**Comment by stsievert** [comment]
Dask clusters can now be used with PyTorch's distributed framework, thanks to the work of Saturn Cloud at https://github.com/saturncloud/dask-pytorch-ddp. This allows use of a Dask cluster with PyTorch's distributed framework (see "[Getting Started with Distributed Data Parallel][1]" for an example). It's similar to the (now archived) [dask-tensorflow]. From their README,

> dask-pytorch-ddp is a Python package that makes it easy to train PyTorch models on Dask clusters using distributed data parallel. The intended scope of the project is
> 
> * bootstrapping PyTorch workers on top of a Dask cluster
> * using distributed data stores (e.g., S3) as normal PyTorch datasets
> * mechanisms for tracking and logging intermediate results, training statistics, and checkpoints.

cc @hhuuggoo and @skirmer.

[1]:https://pytorch.org/tutorials/intermediate/ddp_tutorial.html
[dask-tensorflow]:https://github.com/dask/dask-tensorflow
